crates/sui-core/src/transaction_manager.rs

```diff
@@ -951,28 +951,46 @@ impl TransactionQueue {
         self.digests.is_empty()
     }
 
+    /// Insert the digest into the queue with the given time. If the digest is
+    /// already in the queue, this is a no-op.
     fn insert(&mut self, digest: TransactionDigest, time: Instant) {
-        if self.digests.insert(digest, time).is_none() {
+        if let hash_map::Entry::Vacant(entry) = self.digests.entry(digest) {
+            entry.insert(time);
             self.ages.push((Reverse(time), digest));
         }
     }
 
+    /// Remove the digest from the queue. Returns the time the digest was
+    /// inserted into the queue, if it was present.
+    ///
+    /// After removing the digest, first() will return the new oldest entry
+    /// in the queue (which may be unchanged).
     fn remove(&mut self, digest: &TransactionDigest) -> Option<Instant> {
         let Some(when) = self.digests.remove(digest) else {
             return None;
         };
 
-        while !self.ages.is_empty()
-            && !self
-                .digests
-                .contains_key(&self.ages.peek().expect("heap cannot be empty").1)
-        {
+        // This loop removes all previously inserted entries that no longer
+        // correspond to live entries in self.digests. When the loop terminates,
+        // the top of the heap will be the oldest live entry.
+        // Amortized complexity of `remove` is O(lg(n)).
+        while !self.ages.is_empty() {
+            let first = self.ages.peek().expect("heap cannot be empty");
+
+            // We compare the exact time of the entry, because there may be an
+            // entry in the heap that was previously inserted and removed from
+            // digests, and we want to ignore it. (see test_transaction_queue_remove_in_order)
+            if self.digests.get(&first.1) == Some(&first.0 .0) {
```

Comment:
```suggestion
            if self.digests.get(&first.1) == Some(&first.0.0) {
```

Reply:
That would be nice, but `rustfmt` disagrees